src/tasks/base.py

```diff
@@ -0,0 +1,163 @@
+
+    @abstractmethod
+    def validation_step(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def test_step(self, *args, **kwargs):
+        pass
+
+    def on_train_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def on_test_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def training_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the training epoch with the outputs of all training steps."""
+        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
+        self.log('train/loss', avg_loss, on_step=False, on_epoch=True)
+        self.log('step', self.current_epoch, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('train'))
+
+    def validation_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the validation epoch with the outputs of all validation steps."""
+        avg_loss = torch.stack(outputs).mean()
+        self.log('valid/loss', avg_loss, on_step=False, on_epoch=True)
+        self.log('step', self.current_epoch, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('valid'))
+
+    def test_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of a test epoch with the output of all test steps."""
+        avg_loss = torch.stack(outputs).mean()
+        self.log('test/loss', avg_loss, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('test'))
+
+    def to_onnx(self, onnx_params) -> None:
+        """Is saving the model in ONNX format."""
+        super().to_onnx(input_sample=self.example_input_array,
+                        **onnx_params)
+
+    def on_train_end(self) -> None:
+        """Is calling at the end of training before logger experiment is closed."""
+        onnx_params = self._hparams.onnx_params
+        if onnx_params is not None:
+            self.to_onnx(onnx_params)
+
+    def configure_optimizers(self) -> Tuple[List, List]:
+        """Configure optimizers."""
+        configure_optimizers = self.__constructor.configure_optimizers(self.parameters())
+        optimizers, schedulers = [], []
+
+        for item in configure_optimizers:
+            optimizers.append(item['optimizer'])
+            lr_scheduler = item.get('lr_scheduler')
+            if lr_scheduler is not None:
+                schedulers.append(lr_scheduler['scheduler'])
+            else:
+                schedulers.append(None)
+        return optimizers, schedulers
+
+    def train_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or more PyTorch DataLoaders for training."""
+        data_loader = self.__constructor.create_dataloaders('train')
+        return data_loader
+
+    def val_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for prediction."""
```

Comment:
Why shuffle must be False? Do we really compute on only shuffle False? Also please add documentation